Flaky-test: ConnectionPoolTest.testSetProxyToTargetBrokerAddress

### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.


### Example failure

https://github.com/apache/pulsar/actions/runs/4933518332/jobs/8817664984?pr=20289#step:10:747

### Exception stacktrace

<!-- copy-paste the stack trace in the code block below -->
```
   Error:  org.apache.pulsar.client.impl.ConnectionPoolTest.testSetProxyToTargetBrokerAddress  Time elapsed: 0.085 s  <<< FAILURE!
  java.lang.AssertionError: expected [broker] but found [proxy]
  	at org.testng.Assert.fail(Assert.java:110)
  	at org.testng.Assert.failNotEquals(Assert.java:1577)
  	at org.testng.Assert.assertEqualsImpl(Assert.java:149)
  	at org.testng.Assert.assertEquals(Assert.java:131)
  	at org.testng.Assert.assertEquals(Assert.java:655)
  	at org.testng.Assert.assertEquals(Assert.java:665)
  	at org.apache.pulsar.client.impl.ConnectionPoolTest.testSetProxyToTargetBrokerAddress(ConnectionPoolTest.java:219)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
  	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:139)
  	at org.testng.internal.invokers.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:47)
  	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:76)
  	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
  	at java.base/java.lang.Thread.run(Thread.java:833)
```


### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

--- a/pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ConnectionPoolTest.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ConnectionPoolTest.java
@@ -154,7 +154,7 @@ public class ConnectionPoolTest extends MockedPulsarServiceBaseTest {
     @Test
     public void testSetProxyToTargetBrokerAddress() throws Exception {
         ClientConfigurationData conf = new ClientConfigurationData();
-        conf.setConnectionsPerBroker(5);
+        conf.setConnectionsPerBroker(1);
 
 
         EventLoopGroup eventLoop =
@@ -176,19 +176,24 @@ public class ConnectionPoolTest extends MockedPulsarServiceBaseTest {
             @Override
             protected void doResolveAll(SocketAddress socketAddress, Promise promise) throws Exception {
                 final InetSocketAddress socketAddress1 = (InetSocketAddress) socketAddress;
-                final boolean isProxy = socketAddress1.getHostName().equals("proxy");
-                final boolean isBroker = socketAddress1.getHostName().equals("broker");
+                String hostName = socketAddress1.getHostName();
+                final boolean isProxy = hostName.equals("proxy");
+                final boolean isBroker = hostName.startsWith("broker");
                 if (!isProxy && !isBroker) {
                     promise.setFailure(new IllegalStateException());
                     throw new IllegalStateException();
                 }
                 List<InetSocketAddress> result = new ArrayList<>();
+                // All 127.0.0.0/8 addresses are valid local loopback addresses
                 if (isProxy) {
-                    result.add(new InetSocketAddress("localhost", brokerPort));
-                    result.add(InetSocketAddress.createUnresolved("proxy", brokerPort));
-                } else {
-                    result.add(new InetSocketAddress("127.0.0.1", brokerPort));
-                    result.add(InetSocketAddress.createUnresolved("broker", brokerPort));
+                    result.add(new InetSocketAddress("127.0.0.101", brokerPort));
+                    result.add(new InetSocketAddress("127.0.0.102", brokerPort));
+                } else if (hostName.equals("broker1")){
+                    result.add(new InetSocketAddress("127.0.0.103", brokerPort));
+                    result.add(new InetSocketAddress("127.0.0.104", brokerPort));
+                } else if (hostName.equals("broker2")){
+                    result.add(new InetSocketAddress("127.0.0.105", brokerPort));
+                    result.add(new InetSocketAddress("127.0.0.106", brokerPort));
                 }
                 promise.setSuccess(result);
             }
@@ -203,23 +208,19 @@ public class ConnectionPoolTest extends MockedPulsarServiceBaseTest {
                 InetSocketAddress.createUnresolved("proxy", 9999)).get();
         Assert.assertEquals(cnx.remoteHostName, "proxy");
         Assert.assertNull(cnx.proxyToTargetBrokerAddress);
-        cnx.close();
 
         cnx = pool.getConnection(
-                InetSocketAddress.createUnresolved("broker", 9999),
+                InetSocketAddress.createUnresolved("broker1", 9999),
                 InetSocketAddress.createUnresolved("proxy", 9999)).get();
         Assert.assertEquals(cnx.remoteHostName, "proxy");
-        Assert.assertEquals(cnx.proxyToTargetBrokerAddress, "broker:9999");
-        cnx.close();
+        Assert.assertEquals(cnx.proxyToTargetBrokerAddress, "broker1:9999");
 
 
         cnx = pool.getConnection(
-                InetSocketAddress.createUnresolved("broker", 9999),
-                InetSocketAddress.createUnresolved("broker", 9999)).get();
-        Assert.assertEquals(cnx.remoteHostName, "broker");
+                InetSocketAddress.createUnresolved("broker2", 9999),
+                InetSocketAddress.createUnresolved("broker2", 9999)).get();
+        Assert.assertEquals(cnx.remoteHostName, "broker2");
         Assert.assertNull(cnx.proxyToTargetBrokerAddress);
-        cnx.close();
-
 
         pool.closeAllConnections();
         pool.close();

